staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -683,6 +706,17 @@ func (qs *queueSet) selectQueueLocked() *queue {
 			}
 		}
 	}
+
```

Comment:
I started wondering about picking the queue by "VirtualFinish" time. Up until now, it actually worked fine.

But now, it may become extremely penalizing for wider requests.
Imagine a case where there are just three queue:
- in first two queues you have a number of extremely fast requests with width=1 (constant incoming flow of them so that the queue always have one)
- in the last queue you have a request with say width=2 (but let's say it will also be really fast)
- say we have concurrency limit = 2
Now - because we have G=60s, what is happening is:
queue1: finish_time = t0 + G
queue2: finish_time = t0 + G
queue2: finish_time = to + 2G
[we pick up queue1 and queue2 to dispatch their requests]
let's say they finish after 1ms and we course-correct their virtual time and now we have:
queue1: finish_time = t0 + 1ms + G
queue2: finish_time = t0 + 1ms + G
queue3: finish_time = t0 + 2G
and now we again dispatch from queue1 and queue2

and so on

So until we dispatch something from queue3, it will take ages...


So I think we should actually choose the queue based on VirtualStartTime (as opposed to VirtualFinishTime) and just probably resolve conflicts by preferring narrower requests.

@MikeSpreitzer @tkashem - thoughts?

Reply:
This PR has to be rebased on top of your last PR (this will cause some of your new tests to start failing). 

But instead of changing tests - we should in my opinion address this comment.

Reply:
FWIW - for requests with width=1, this would be a no-op change effectively

Reply:
so i tracked how `queue.virtualStart` changes in order of the following events:

A: a new request arrives (enqueue)
- if the queue is inactive (no request is currently executing or no other request is in queue) then new request can start immediately (current `R(t)` value):
```
queue.virtualStart = qs.virtualTime
```
- if the queue is active then we don't change `queue.virtualStart`

B: next round - select a queue with the minimum finish time
- we make sure per-queue virtual time `queue.virtualStart` does not fall behind the global

C: the request is dequeued, we adjust `queue.virtualStart` as below
```
queue.virtualStart += G*W
```

D: request finishes - we do course correcion
```
r.queue.virtualStart -= (G-S)*W
```

the scenario you described, yes it would be a while before q3 is picked and that's fair i think if we consider the bit by bit round robin, at some point virtual finish time for q3 will be minimum


 

If we consider the the above operations, it seems we do take into account the arrival time and implement
```
Si = MAX(R(t), Fi-1)
```

Reply:
At some point it *would* be minimal I agree. The problem is huge how G is (it's 60s).
So given we're multiplying it by `seats`, it may take minutes until it would happen. Up until then we wll be course-correcting virtualStart (point D) that will still be pushing other queues to have smaller time.

And because of that, at some point we will simply reject that calls (as it would be waiting for too old).

Which makes this unfair, because once request with high width reaches the peak of queue, we effectively stop choosing that queue for a long time of there are requests with smaller width in other queue (assuming the wide request didn't arrive minutes earlier).

Reply:
In other words - if we assume that processing each request takes 1ms (which isn't completely unrealistic), then you have to process 60,000 requests from queue1 and queue2 before you pick the request from queue3. Given that the second request has just 2x higher width, that's not fair at all.

I strongly believe that just using "virtual_start" is the right thing to do:
- for requests with width=1, it's a no-op (because finish_time = start_time + G so finish_time_1 < finish_time_2 <=> start_time_1 < start_time_2)
- it achieves what we want for requests with higher width

Reply:
so let's simplify the scenario you described:
- we have two queues `q1` and `q2`
- `q1` has an infinite supply of requests with `W=1`
- `q2` has one request waiting in the queue with the same width `W` = 1
- requests complete really fast `S=1ms` 

In this scenario we will execute 60,000 requests from `q1` before we pick the request from `q2`. since we are adjusting the virtual start for the queue it may look like we are being unfair to `q2` with a request of same width. Yes it gets proportionally worse for request with `W` > 1. 


So what you are suggesting would be equivalent to what we have today, not take into account `W` into virtual finish time calculation (basically `W` = 1 for all requests). For now, if we don't use `W` in calculating virtual finish time then we are retaining the current behavior. This feels safer, at least for now.
I think we need unit tests where we can simulate these use cases in a controlled manner and observer/assert on the expected behavior with multiple queues and requests with varying width. This would require some refactor of the `queueSet` and carefully crafted unit tests.

In the meantime, we could have the following:
- do not take `W` into account when calculating virtual finish time - in the virtual world all requests are estimated to take the same amount of time.
- we use `W` for seats allocation.

@wojtek-t @MikeSpreitzer thoughts?

Reply:
> q2 has one request waiting in the queue with the same width W = 1

You mean q2 has request with w=2, right?

> - do not take W into account when calculating virtual finish time - in the virtual world all requests are estimated to take the same amount of time.
>  - we use W for seats allocation.

So to rephrase:
- we still use W to adjusting virtualStart as we do now, e.g. here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go#L640
- but we just change the GetNextFinish() to not take `W` into account

That translates to exactly the same what I'm proposing (i.e. taking virtualStart, because virtualFinish for every request will be equal to "virtualStart + G". So I don't really see how this is different from what I'm proposing.

Reply:
> You mean q2 has request with w=2, right?

no, q2 has request with w=1 but I forgot to mention concurrency limit = 1

> That translates to exactly the same what I'm proposing

I was referring to:
- A: when we dequeue: `queue.virtualStart += G`
- B: when request finishes: `queue.virtualStart -= G-S`
- C: when we pick: finish time `queue.virtualStart + G`

This is the current behavior, what you are suggesting is for `C` we don't use `W`, but otherwise use it for A and B. with your proposal, while there is a currently executing request with much higher width, smaller requests waiting in the said queue won't be picked up since its `virtualStart` is higher because we included `W` and these requests waiting have a chance to timeout as well. 
with my proposal we are equally fair to all requests, `W` matters in only seat allocation. it's practically no change from how we select from queue today. 

we can start with what you are proposing, we can tweak later if need be. @MikeSpreitzer unless you have any other concerns i want to go ahead and make the changes

Reply:
> no, q2 has request with w=1 but I forgot to mention concurrency limit = 1

With width=1 everything works correctly. Because virtualStart will be increasing after every request so it doesn't matter.

> This is the current behavior

Current where? The current code behavior is already that we increase virtualStart by G*Seats() - see link I pasted above:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go#L640

> with your proposal, while there is a currently executing request with much higher width, smaller requests waiting in the said queue won't be picked up since its virtualStart is higher because we included W and these requests waiting have a chance to timeout as well.

Which is exactly what I want. We need to be fair across queue - not fair across requests.
This has to be adressed by changing how we choose to which queue we add a request - see:
https://github.com/kubernetes/kubernetes/pull/102875/files#r653296345


> with my proposal we are equally fair to all requests, W matters in only seat allocation. it's practically no change from how we select from queue today.

But that's not what we want.
Imagine two queues:
- q1 we have a number of short 'get object X' requests (width=1)
- q2 we have a number of large 'get all pods" requests (e.g. width=10)

What we really want is we want to provide equal capacity across these queue. So if the request is e.g. 10x more expensive we want to proceed 10x less requests from that queue.

So to summarize - we want to be fair across queues. What you're proposing doesn't provide that guarantee - wide requests to some extent will be starving small short requests, which is not what we want.
What I'm proposing provide the fairness across queues.